Cyvcf2 dumpstr

LiterallyUniqueLogin opened the pull request:
Please finish cyvcf2 qcSTR PR before this.

DumpSTR now runs on cyvcf2. See RELEASE_NOTES.txt for a description of the changes. I'll be adding more to that file for the other cyvcf2 PRs (both the ones to come and the previously released ones which should be documented)

Only piece of info not in that file that comes to mind is that I commented out the code in tests/test_filters.py as I did not update it to the new syntax. All that code it covered is still tested I believe, but less concretely. If someone wants to do that that'd be more than welcome.

nmmsv commented on trtools/dumpSTR/tests/test_dumpSTR.py:
```diff
@@ -58,18 +60,51 @@ def args(tmpdir):
+    assert ("FILTER" in captured.err and "AC" in captured.err
+            and "REFAC" in captured.err)
+
+def test_WorrisomePreexistingFilter(args, testDumpSTRdir, capsys):
+    fname = os.path.join(testDumpSTRdir, "worrisome_preexisting_filter.vcf")
```
Can you maybe add a couple of lines of comment for each test to say what's being tested?

LiterallyUniqueLogin replied:
I added comments to all the new tests.

nmmsv commented on trtools/dumpSTR/tests/test_dumpSTR.py:
```diff
@@ -58,18 +60,51 @@ def args(tmpdir):
+    retcode = main(args)
+    assert retcode == 0
+    captured = capsys.readouterr()
+    assert 'HWE0.5' in captured.err
```
I think having float values in the filter name is probably not a good idea. We will have a zillion different filter values :D 
I think this is another reason not to include the value with the filter names.

LiterallyUniqueLogin replied:
This was old behavior for locus level filters, which is what you're seeing here, not a change I made. See the email chain.

LiterallyUniqueLogin replied:
Also, to clarify: the value there is the threshold, not the filtered value, so there'll only be one ID for each locus-level filter. (And no IDs for the call-level filters, just ##FORMAT=<ID="FITLER", ...> )

nmmsv commented on trtools/dumpSTR/tests/test_dumpSTR.py:
```diff
@@ -58,18 +60,51 @@ def args(tmpdir):
 # Test if basic inputs and threshold filters work for each file
-def test_GangSTRFile(args, vcfdir):
-    fname = os.path.join(vcfdir, "test_gangstr.vcf")
+def test_GangSTRFile(args, testDumpSTRdir):
+    fname = os.path.join(testDumpSTRdir, "trio_chr21_gangstr.sorted.vcf.gz")
     args.vcf = fname
```
I think the tests under these functions only check if the code works without error. Unless I'm missing something, I think we need to tests under test_filters.py for the actual functionality of each filter that is implemented.

nmmsv commented on trtools/testsupport/sample_vcfs/dumpSTR_vcfs/create_test_files.sh:
```diff
@@ -0,0 +1,38 @@
+# Note: this command should be run from the root of the repo
+
+# Note: the thresholds in this file are arbitrary and just used for testing
+# do not assume they are reasonable for data analysis
+
```
I think this is a really cool idea! Thanks for implementing it!

LiterallyUniqueLogin replied:
Yeah! This took me quite a bit of effort, but it also caught a bunch of unintended changes that I made when I swapped the code over. So feels worth it.

nmmsv commented on trtools/dumpSTR/filters.py:
```diff
@@ -440,21 +488,42 @@ class HipSTRCallMinSuppReads(Reason):
-        min_read_count = min([r1, r2])
-        if min_read_count < self.threshold: return min_read_count
-        else: return None
+        # Going to assume that either all samples are phased or none are
```
I don't think this assumption necessarily holds. Not sure though, you have worked with HipSTR data more than I.

LiterallyUniqueLogin replied:
Not sure if this is a reasonable assumption, I agree its a bit sketchy. Its what was already being done though, so I'm viewing it as an issue that isn't being fixed which I'm okay with as opposed to an issue I'm introducing.

LiterallyUniqueLogin replied:
Honestly, none of our code is sensitive to phasing. I'm not sure where else that is an issue, but it isn't being tested for at all. Handling indeterminate phasing properly would probably involve a lot of testing in a lot of places.

nmmsv commented on trtools/dumpSTR/filters.py:
```diff
@@ -592,25 +697,38 @@ class GangSTRCallBadCI(Reason):
+        return sample_filter
+
+'''
+Since this class is experimental, I'm just not converting it from pyvcf to
+cyvcf2 for now
+class GangSTRCallRequireSupport(Reason)
```
I think this class (and this flag) is used for Ileena's paper and we probably have to use it again for future projects.

LiterallyUniqueLogin replied:
I'm trying to get these types of side projects done quicker. I'd prefer to push this now and then get back to this when it's needed.

nmmsv reviewed (changes requested):
Wow this must have been a lot of work! Thanks for undertaking the huge challenge. I think in general everything looks reasonable. I have added some comments too. I think the main points that I'm worried about are:
* tests for each filter separately: We have tests to check errors are not thrown, and we have tests to compare final VCFs, but I think what happens in the filter classes in between isn't tested.
* require support filter for GangSTR: This is something that we have used before and probably would use again. It would be good to get Melissa's input on it too.

That's all that I remember now, I'll add if I remembered something else.
Thanks again for the hard work!

LiterallyUniqueLogin commented on RELEASE_NOTES.rst:
```diff
@@ -5,12 +5,31 @@ Features:
+  will overwrite them and issue a warning in case that was not intended)
+
+Output changes:
+* DumpSTR call level filters now have the value of the call which triggered
+  the filter appended to the filter name in the FILTER format field. (e.g.
+  GangSTRCallMinDepth12 because the field had a depth of 12 and that's lower
+  than the required min depth)
```
We were calculating all these values before and then throwing them away, I assumed they were for some reason and so included them. But perhaps not. I'll ping Melissa and see what she thinks.

LiterallyUniqueLogin commented on RELEASE_NOTES.rst:
```diff
@@ -27,6 +46,26 @@ Python interface changes:
+
+Regressions:
+
+* The --gangstr-require-support filter has been disabled.
```
Nope. The next user would have to update it.

nmmsv replied:
I think we should include this before merging as it's something that is used on almost all gangstr filtering jobs that we have. I can work on it if you don't have time.

nmmsv replied:
On second thought, I can do this later. I need to spend some time on it and add tests and such, which can take some time.

LiterallyUniqueLogin commented on trtools/dumpSTR/dumpSTR.py:
```diff
@@ -514,102 +524,167 @@ def GetAllCallFilters(call_filters):
-    call_filters : list of filters.Reason
+    record :
+       The record to apply filters to.
+       Note: once this method has been run, this object
+       will be in an inconsistent state. All further use
+       should be directed towards the returned TRRecord object.
```
That'd be a fair bit of hassle - the TRRecord constructor is quite particular so cloning it would have to have a lot of checks to figure out which optional arguments to pass through. The idea is a good one, I'd do it if it weren't a pain, but given that the code works as is and we don't expect end users to call this function I think I'd like to pass on it.

LiterallyUniqueLogin replied:
Rethinking my response:

The issue is that it's not the trh.TRRecord object that's being modified but rather the cyvcf2.Variant object that it contains, which is a necessary step in using cyvcf2.Writer. The appropriate thing to do would be to clone the underlying cyvcf2.Variant object and then work with the clone, which is a good idea and not something I thought of. Unfortunately, I just checked and that cyvcf2.Variant doesn't have a clone method. So this is a no go - we have to modify the original cyvcf2.Variant, and that necessarily puts the original wrapping TRRecord in an inconsistent state.

The only other fix would be to modify the original wrapping TRRecord to fix it, but that seems like it would be very error-prone code. Creating a new TRRecord that's bound to be in a clean state seems better, and that's what the code currently does.

nmmsv replied:
ok that's fair. I thought it would be simple, but it it's very involved, we don't need to do it. It's unlikely that we need the original object down the line anyway.

LiterallyUniqueLogin commented on trtools/dumpSTR/dumpSTR.py:
```diff
@@ -514,102 +524,167 @@ def GetAllCallFilters(call_filters):
+        if np.all(nans):
+            continue
+        sample_info[filt.name] += np.logical_and(~nans, ~nocalls)
+        filt_output_text = np.char.mod('%g', filt_output)
+        filt_output_text = np.char.add(filt.name, filt_output_text)
+        filt_output_text[nans] = ''
+        not_first_filter = np.logical_and(~nans, all_filter_text != '')
+        all_filter_text[not_first_filter] = \
+            np.char.add(all_filter_text[not_first_filter], ',')
+        all_filter_text = np.char.add(all_filter_text, filt_output_text)
```
Adding!

LiterallyUniqueLogin commented on trtools/dumpSTR/dumpSTR.py:
```diff
@@ -514,102 +524,167 @@ def GetAllCallFilters(call_filters):
+    if np.any(nocalls):
+        nocall_text = np.empty((nocalls.shape[0]), dtype='U6')
+        nocall_text[nocalls] = 'NOCALL'
+        # if there was already a no call, don't say we applied
+        # our filters to it
```
Just to make clear what currently happens: Take this example as input to DumpSTR
GT:Q     .|.:5
Since it's already filtered, I had DumpSTR emit
GT:Q:FILTER     .|.:5:
and not
GT:Q:FILTER     .|.:.:NOCALL

That way it's clear what DumpSTR filtered and what it didn't. But I'm not sold on that choice. Do you prefer the alternative?

nmmsv replied:
humm I think the Filter reason is missing on the example that you gave (for what you currently emit). I think if the FILTER field is incluced, it should be better than the last one.

LiterallyUniqueLogin replied:
Sorry, my understanding/explanation above was just wrong. If a field was already filtered, then we append "NOCALL" to it but don't append any other filter reasons. E.g. if we have --min-q 20,
GT:Q .|.:5
becomes
GT:Q:FILTER .|.:5:NOCALL
and not
GT:Q:FILTER .|.:5:MINQ20_5,NOCALL

LiterallyUniqueLogin commented on RELEASE_NOTES.rst:
```diff
@@ -5,12 +5,31 @@ Features:
+  samplog output file
+* DumpSTR now adds ##FILTER=<ID=PASS,Description="All filters passed">
+  to the header line
+* When DumpSTR filters a call and replaces each of its format fields with the no call
+  '.', fields with number>1 are now replaced with '.,.' e.g. for a number=2 field, rather
+  than just a single '.'
```
Changed

LiterallyUniqueLogin commented on RELEASE_NOTES.rst:
```diff
@@ -5,12 +5,31 @@ Features:
 * Underlying libraries now use cyvcf2 instead of PyVCF for VCF parsing.
   This makes both the underlying VCF reading code and the TRTools code
-  significantly faster and more memory efficient. For instance, the loading of
+  significantly faster and memory efficient. For instance, the loading of 
```
Changed

LiterallyUniqueLogin commented on trtools/dumpSTR/dumpSTR.py:
```diff
@@ -514,102 +524,167 @@ def GetAllCallFilters(call_filters):
+        record.vcfrecord.set_format(field, vals)
+
+    # rebuild the TRRecord with the newly modified cyvcf2 vcfrecord
+    # TODO test that this works under all permuations of inputs
```
Good catch.

I just ran code coverage, these branches are all covered (presumably by the test_output tests) except for the record.HasFabricatedRefAllele() section. That's not tested because I don't have an ExpansionHunter test because the EH filters still don't work. When someone fixes those and adds a test for it that'll get covered. As of now the rest of the lines there are tested and working!

LiterallyUniqueLogin commented:
Hey Nima,

Thank you for doing a thorough review! I've added comments where asked, and added filter level tests for the locus-level filters and for the HipSTR call-level filters. Per the email chain I'll change the call-level filters from MinQ12 to e.g. MinQ30_12 which hopefully is slightly easier for parsing due to the underscore.

You can check the diff of what I changed since your review [here](https://github.com/gymreklab/TRTools/compare/0bafb04...cyvcf2_dumpstr).

I'm still intending to punt on the GangSTR filter and the other individual filter tests - not arguing that they're not worthwhile, but the code is currently tested better than when I found it, and the test comparing the output to the previous iterations has shown me that the filters work at least as well as they did before. I or someone else can come back to implement them when its the best use of that person's time.

nmmsv commented:
Thanks for the updates! I think we can add the additional pieces (the rest of filter tests and gangstr require support) after everything is merged.
